File: Controllers/OrderController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Identity;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using Sofranet.Data;
using Sofranet.Helpers;
using Sofranet.Models;
using Sofranet.Services;

namespace Sofranet.Controllers;

[Authorize]
public class OrderController : Controller
{
    private readonly ApplicationDbContext _db;
    private readonly UserManager<ApplicationUser> _userManager;
    private readonly IEmailService _email;
    private readonly IPdfService _pdf;
    private readonly ILogService _logs;

    private const string CART_KEY = "cart";
    private const string PAYMENT_KEY = "payment_approved";
    private const string DELIVERY_KEY = "delivery_info";
    private const string PENDING_KEY = "cart_pending";

    public OrderController(ApplicationDbContext db,
        UserManager<ApplicationUser> userManager,
        IEmailService email,
        IPdfService pdf,
        ILogService logs)
    {
        _db = db;
        _userManager = userManager;
        _email = email;
        _pdf = pdf;
        _logs = logs;
    }

    [HttpPost]
    [ValidateAntiForgeryToken]
    [Authorize(Roles = "User")]
    public async Task<IActionResult> Create()
    {
        if (HttpContext.Session.GetString(PAYMENT_KEY) != "true")
        {
            TempData["ErrorMessage"] = "Önce ödeme onayı gerekli";
            return RedirectToAction("Index", "Cart");
        }

        var cart = HttpContext.Session.GetObject<List<CartItem>>(CART_KEY);
        if (cart == null || cart.Count == 0)
        {
            TempData["ErrorMessage"] = "Sepet boş";
            return RedirectToAction("Index", "Menu");
        }

        var delivery = HttpContext.Session.GetObject<DeliveryInfo>(DELIVERY_KEY);
        if (delivery == null)
        {
            TempData["ErrorMessage"] = "Teslimat bilgisi eksik";
            return RedirectToAction("Index", "Cart");
        }

        var user = await _userManager.GetUserAsync(User);
        if (user == null) return RedirectToAction("Login", "Account");

        // Enrich - fiyat hesabı
        var menuIds = cart.Select(c => c.MenuItemId).Distinct().ToList();
        var optIds = cart.SelectMany(c => c.OptionIds).Distinct().ToList();

        var menuMap = await _db.MenuItems
            .Where(m => menuIds.Contains(m.Id))
            .ToDictionaryAsync(m => m.Id);
        var optMap = await _db.MenuOptions
            .Where(o => optIds.Contains(o.Id))
            .ToDictionaryAsync(o => o.Id);

        decimal total = 0;
        foreach (var c in cart)
        {
            if (!menuMap.TryGetValue(c.MenuItemId, out var mi)) continue;
            decimal extras = c.OptionIds.Sum(oid => optMap.TryGetValue(oid, out var o) ? o.ExtraPrice : 0);
            total += (mi.Price + extras) * c.Quantity;
        }

        using var tx = await _db.Database.BeginTransactionAsync();
        int orderId;
        try
        {
            var order = new Order
            {
                UserId = user.Id,
                CatererId = cart[0].CatererId,
                TotalAmount = total,
                Status = "pending",
                PaymentStatus = "paid",
                DeliveryAddress = delivery.Address,
                DeliveryLat = delivery.Lat,
                DeliveryLng = delivery.Lng,
                CreatedAt = DateTime.UtcNow
            };
            _db.Orders.Add(order);
            await _db.SaveChangesAsync();

            foreach (var c in cart)
            {
                if (!menuMap.TryGetValue(c.MenuItemId, out var mi)) continue;
                decimal extras = c.OptionIds.Sum(oid => optMap.TryGetValue(oid, out var o) ? o.ExtraPrice : 0);

                var orderItem = new OrderItem
                {
                    OrderId = order.Id,
                    MenuItemId = c.MenuItemId,
                    Quantity = c.Quantity,
                    ItemPrice = mi.Price,
                    CustomizationExtra = extras
                };
                _db.OrderItems.Add(orderItem);
                await _db.SaveChangesAsync();

                foreach (var oid in c.OptionIds)
                {
                    _db.OrderItemOptions.Add(new OrderItemOption
                    {
                        OrderItemId = orderItem.Id,
                        OptionId = oid
                    });
                }
                foreach (var rid in c.RemovalIds)
                {
                    _db.OrderItemRemovals.Add(new OrderItemRemoval
                    {
                        OrderItemId = orderItem.Id,
                        RemovableIngredientId = rid
                    });
                }
            }
            await _db.SaveChangesAsync();
            await tx.CommitAsync();
            orderId = order.Id;
        }
        catch (Exception ex)
        {
            await tx.RollbackAsync();
            await _logs.LogAsync(HttpContext, "ERROR", $"order_create_failed err={ex.Message}", user.Id);
            TempData["ErrorMessage"] = "Sipariş oluşturulurken hata oluştu";
            return RedirectToAction("Index", "Cart");
        }

        // Sepeti temizle
        HttpContext.Session.Remove(CART_KEY);
        HttpContext.Session.Remove(PAYMENT_KEY);
        HttpContext.Session.Remove(DELIVERY_KEY);
        HttpContext.Session.Remove(PENDING_KEY);

        await _logs.LogAsync(HttpContext, "ORDER_CREATED",
            $"order_id={orderId} total={total}", user.Id);

        // Email - hata olursa akış kırılmasın
        try
        {
            var info = await OrderHelper.FetchOrderFullAsync(_db, orderId);
            if (info != null)
            {
                var userMail = await _email.SendOrderConfirmationToUserAsync(info);
                var catererMail = await _email.SendOrderNotificationToCatererAsync(info);
                if (userMail || catererMail)
                {
                    await _logs.LogAsync(HttpContext, "EMAIL_SENT",
                        $"order_id={orderId} user_ok={userMail} caterer_ok={catererMail}", user.Id);
                }
                else
                {
                    await _logs.LogAsync(HttpContext, "EMAIL_FAILED",
                        $"order_id={orderId} smtp_not_configured", user.Id);
                }
            }
        }
        catch (Exception ex)
        {
            await _logs.LogAsync(HttpContext, "EMAIL_FAILED",
                $"order_id={orderId} err={ex.Message}", user.Id);
        }

        TempData["SuccessMessage"] = "Siparişin oluşturuldu!";
        return RedirectToAction(nameof(Success), new { id = orderId });
    }

    [Authorize(Roles = "User")]
    public async Task<IActionResult> Success(int id)
    {
        var info = await OrderHelper.FetchOrderFullAsync(_db, id);
        if (info == null) return NotFound();

        var userId = _userManager.GetUserId(User);
        if (info.Order.UserId != userId)
        {
            TempData["ErrorMessage"] = "Bu sipariş sana ait değil";
            return RedirectToAction(nameof(MyOrders));
        }
        ViewData["Title"] = $"Sipariş #{info.Order.Id}";
        return View(info);
    }

    [Authorize(Roles = "User")]
    public async Task<IActionResult> MyOrders(string? status, int page = 1)
    {
        if (page < 1) page = 1;
        var userId = _userManager.GetUserId(User);
        var query = _db.Orders.Where(o => o.UserId == userId);
        if (!string.IsNullOrEmpty(status))
            query = query.Where(o => o.Status == status);

        var total = await query.CountAsync();
        var orders = await query
            .Include(o => o.Caterer)
            .OrderByDescending(o => o.CreatedAt)
            .Skip((page - 1) * 15).Take(15)
            .Select(o => new OrderListItem
            {
                Id = o.Id,
                TotalAmount = o.TotalAmount,
                Status = o.Status,
                CreatedAt = o.CreatedAt,
                CompletedAt = o.CompletedAt,
                CatererName = o.Caterer!.FullName,
                ItemCount = _db.OrderItems.Count(i => i.OrderId == o.Id),
                HasRating = _db.Ratings.Any(r => r.OrderId == o.Id)
            })
            .ToListAsync();

        ViewBag.Status = status;
        ViewBag.CurrentPage = page;
        ViewBag.TotalPages = (int)Math.Ceiling(total / 15.0);
        ViewBag.Total = total;
        ViewData["Title"] = "Siparişlerim";
        return View(orders);
    }

    // === PDF ===
    public async Task<IActionResult> Receipt(int id)
    {
        var info = await OrderHelper.FetchOrderFullAsync(_db, id);
        if (info == null) return NotFound();
        if (!await CanAccessOrder(info)) return Forbid();

        var bytes = _pdf.GenerateReceipt(info);
        return File(bytes, "application/pdf", $"sofranet-makbuz-{id}.pdf");
    }

    public async Task<IActionResult> Agreement(int id)
    {
        var info = await OrderHelper.FetchOrderFullAsync(_db, id);
        if (info == null) return NotFound();
        if (!await CanAccessOrder(info)) return Forbid();

        var bytes = _pdf.GenerateAgreement(info);
        return File(bytes, "application/pdf", $"sofranet-sozlesme-{id}.pdf");
    }

    private Task<bool> CanAccessOrder(FullOrderInfo info)
    {
        var userId = _userManager.GetUserId(User);
        if (User.IsInRole("Admin")) return Task.FromResult(true);
        if (User.IsInRole("User")) return Task.FromResult(info.Order.UserId == userId);
        if (User.IsInRole("Caterer")) return Task.FromResult(info.Order.CatererId == userId);
        return Task.FromResult(false);
    }
}

public class OrderListItem
{
    public int Id { get; set; }
    public decimal TotalAmount { get; set; }
    public string Status { get; set; } = "";
    public DateTime CreatedAt { get; set; }
    public DateTime? CompletedAt { get; set; }
    public string CatererName { get; set; } = "";
    public int ItemCount { get; set; }
    public bool HasRating { get; set; }
}

File: controllers/orderController.js
// order controller - siparis olusturma + listeleme
const db = require('../config/db');
const { enrichCart } = require('../utils/cartHelper');
const { fetchOrderFull } = require('../utils/orderHelper');
const emailService = require('../services/emailService');
const pdfService = require('../services/pdfService');

async function create(req, res) {
  // payment + sepet kontrolu
  if (!req.session.paymentApproved) {
    req.flash('error', 'Önce ödemeyi tamamlamalısın');
    return res.redirect('/cart/checkout');
  }
  if (!req.session.cart || req.session.cart.length === 0) {
    req.flash('error', 'Sepetin boş');
    return res.redirect('/menu');
  }

  let client;
  try {
    const cart = req.session.cart;
    const { cartItems, total } = await enrichCart(cart);
    const userId = req.session.user.id;
    const catererId = cart[0].catererId;
    const delivery = req.session.deliveryInfo || {};

    client = await db.getClient();
    await client.query('BEGIN');

    const orderRes = await client.query(
      `INSERT INTO orders
        (user_id, caterer_id, total_amount, status, payment_status, delivery_address, delivery_lat, delivery_lng)
       VALUES ($1, $2, $3, 'pending', 'paid', $4, $5, $6)
       RETURNING id`,
      [userId, catererId, total, delivery.address || null, delivery.lat || null, delivery.lng || null]
    );
    const orderId = orderRes.rows[0].id;

    for (const ci of cartItems) {
      const oiRes = await client.query(
        `INSERT INTO order_items (order_id, menu_item_id, quantity, item_price, customization_extra)
         VALUES ($1, $2, $3, $4, $5)
         RETURNING id`,
        [orderId, ci.menuItem.id, ci.quantity, ci.basePrice, ci.extras]
      );
      const orderItemId = oiRes.rows[0].id;

      for (const o of ci.options) {
        await client.query(
          'INSERT INTO order_item_options (order_item_id, option_id) VALUES ($1, $2)',
          [orderItemId, o.id]
        );
      }
      for (const r of ci.removals) {
        await client.query(
          'INSERT INTO order_item_removals (order_item_id, removable_ingredient_id) VALUES ($1, $2)',
          [orderItemId, r.id]
        );
      }
    }

    await client.query('COMMIT');

    // sepeti ve payment state'i temizle
    req.session.cart = [];
    delete req.session.paymentApproved;
    delete req.session.deliveryInfo;
    delete req.session.pendingItem;

    // email gonderimi - user ve caterer bilgilerini cek
    try {
      const usrRes = await db.query(
        'SELECT name, email, phone FROM users WHERE id = $1',
        [userId]
      );
      const catRes = await db.query(
        'SELECT name, email, phone FROM users WHERE id = $1',
        [catererId]
      );
      const user = usrRes.rows[0];
      const caterer = catRes.rows[0];

      const orderObj = {
        id: orderId,
        total_amount: total,
        delivery_address: delivery.address || null,
        created_at: new Date()
      };

      // cartItems zaten enrich edilmis (basePrice, extras, options, removals dolu)
      await emailService.sendOrderConfirmationToUser(user, orderObj, cartItems, caterer);
      await emailService.sendOrderNotificationToCaterer(caterer, user, orderObj, cartItems);
      console.log('LOG: EMAIL_SENT user=', user.email, 'caterer=', caterer.email, 'order=', orderId);
    } catch (e) {
      console.error('email gonderim hata (order proceeds):', e.message);
    }

    // TODO: Faz 7 - PDF generation
    console.log('TODO: Generate receipt and agreement PDFs for order', orderId);
    // TODO: Faz 10 - logging tablosuna yaz
    console.log('LOG: ORDER_CREATED user=', userId, 'order=', orderId);

    res.redirect('/orders/' + orderId + '/success');
  } catch (err) {
    if (client) {
      try { await client.query('ROLLBACK'); } catch (e) { console.error('rollback hata:', e.message); }
    }
    console.error('order create error:', err);
    req.flash('error', 'Sipariş oluşturulurken hata olustu');
    res.redirect('/cart/checkout');
  } finally {
    if (client) client.release();
  }
}

// ownership: user kendi siparisi / caterer kendi siparisi / admin
function canAccessOrder(order, sessionUser) {
  if (!sessionUser || !order) return false;
  if (sessionUser.role === 'admin') return true;
  if (sessionUser.role === 'user' && order.user_id === sessionUser.id) return true;
  if (sessionUser.role === 'caterer' && order.caterer_id === sessionUser.id) return true;
  return false;
}

async function successPage(req, res) {
  try {
    const id = parseInt(req.params.id, 10);
    if (isNaN(id)) return res.status(404).send('Sipariş bulunamadı');

    const data = await fetchOrderFull(id);
    if (!data || data.order.user_id !== req.session.user.id) {
      return res.status(404).send('Sipariş bulunamadı');
    }

    // success view'inin kullandigi item format orderHelper.js'tekiyle uyumlu olsun
    // (item.name, item.options[].group_name+name+extra_price, item.removals[].name, item_price, customization_extra)
    const items = data.items.map((it) => ({
      id: it.id,
      menu_name: it.name,
      quantity: it.quantity,
      item_price: it.basePrice,
      customization_extra: it.customizationExtra,
      options: it.options.map(o => ({
        group_name: o.groupName,
        name: o.name,
        extra_price: o.extra
      })),
      removals: it.removals
    }));

    res.render('user/order-success', {
      title: 'Sipariş #' + id,
      order: { ...data.order, caterer_name: data.caterer.name },
      items
    });
  } catch (err) {
    console.error('order success error:', err);
    res.status(500).send('Sipariş yüklenemedi');
  }
}

async function downloadReceipt(req, res) {
  try {
    const id = parseInt(req.params.id, 10);
    if (isNaN(id)) return res.status(404).send('Sipariş bulunamadi');

    const data = await fetchOrderFull(id);
    if (!data || !canAccessOrder(data.order, req.session.user)) {
      return res.status(404).send('Sipariş bulunamadi');
    }

    const buffer = await pdfService.generateReceipt(data.order, data.items, data.user, data.caterer);

    res.setHeader('Content-Type', 'application/pdf');
    res.setHeader('Content-Disposition', `attachment; filename="sofranet-makbuz-${id}.pdf"`);
    res.send(buffer);
  } catch (err) {
    console.error('receipt download error:', err);
    res.status(500).send('Makbuz uretilemedi');
  }
}

async function downloadAgreement(req, res) {
  try {
    const id = parseInt(req.params.id, 10);
    if (isNaN(id)) return res.status(404).send('Sipariş bulunamadi');

    const data = await fetchOrderFull(id);
    if (!data || !canAccessOrder(data.order, req.session.user)) {
      return res.status(404).send('Sipariş bulunamadi');
    }

    const buffer = await pdfService.generateAgreement(data.order, data.items, data.user, data.caterer);

    res.setHeader('Content-Type', 'application/pdf');
    res.setHeader('Content-Disposition', `attachment; filename="sofranet-sozlesme-${id}.pdf"`);
    res.send(buffer);
  } catch (err) {
    console.error('agreement download error:', err);
    res.status(500).send('Sözleşme üretilemedi');
  }
}

async function myOrders(req, res) {
  try {
    const result = await db.query(
      `SELECT o.id, o.total_amount, o.status, o.created_at, o.completed_at,
        u.name as caterer_name,
        (SELECT COUNT(*) FROM order_items WHERE order_id = o.id) as item_count,
        EXISTS (SELECT 1 FROM ratings WHERE order_id = o.id) as has_rating
       FROM orders o
       JOIN users u ON u.id = o.caterer_id
       WHERE o.user_id = $1
       ORDER BY o.created_at DESC`,
      [req.session.user.id]
    );
    res.render('user/orders', {
      title: 'Siparişlerim',
      orders: result.rows
    });
  } catch (err) {
    console.error('my orders error:', err);
    res.status(500).send('Siparişler yüklenemedi');
  }
}

module.exports = { create, successPage, myOrders, downloadReceipt, downloadAgreement };
